pkg/operator/configobservation/etcd/observe_etcd.go

```diff
@@ -190,10 +191,30 @@ func ObservePendingClusterMembers(genericListers configobserver.Listers, recorde
 	return observedConfig, nil
 }
 
-func isPendingReady(bucket string, podName string, scalingName string) bool {
+func isPendingReady(bucket string, podName string, scalingName string, podLister corelistersv1.PodLister) bool {
 	if bucket == "pending" && podName != scalingName {
 		return false
 	}
+	pod, err := podLister.Pods(clustermembercontroller.EtcdEndpointNamespace).Get(podName)
+	if err != nil {
+		klog.Errorf("isPendingReady: error getting pod %#v", err)
+		return false
+	}
+	if len(pod.Status.InitContainerStatuses) < 2 {
+		klog.Infof("isPendingRead: waiting for init cert containers to pass")
+		return false
+	} else {
+		if pod.Status.InitContainerStatuses[1].State.Terminated != nil && pod.Status.InitContainerStatuses[1].State.Terminated.ExitCode == 0 {
+			if pod.Status.ContainerStatuses[0].State.Terminated != nil && pod.Status.ContainerStatuses[0].RestartCount > 0 {
+				klog.Info("isPendingRead: restart count >0, Remove state")
```

Comment:
This makes me feel that we could become wedged if an initcontainer[0] for unknown reasons has a restart. Why do we care if discovery has a restart but then exit 0? How does Pod recover if `pod.Status.ContainerStatuses[0].RestartCount > 0`

Reply:
Good point. 

For instance, the discovery container could fail during bootstrap because dns in not set up. But in that case, we loop through to wait for dns to not really failing exiting with 0. Looking at discovery code, is there a way to make reasonable assumptions that discovery will never exit or exit 0? 

Also, if discovery indeed exits say a non 0 number, it will be restarted by kube until it exits 0 before certs container starts. If 
`pod.Status.InitContainerStatuses[1].State.Terminated.ExitCode == 0`, my understanding is that it would be safe to assume discovery exit code is 0.

and lastly my assumption was that after the restart controller takes the pod out from remove state, it RestartCount will be set to 0, so it will still recover. I can see now how this assumption might be wrong.

Reply:
I think we need to wire something up for multiple scenarios:

Scenario 1: Bootstrapping, member is in PendingUnknown State
Scenario 2: Restart, member is in PendingRemove State

I suggest the following: 

1. If the member is in PendingUnknown State, and if we see two containers terminated with exit code 0, return true
2. If the member is in RemoveState and if we see two containers terminated return true.

The fact that after `Remove`, restart controller will kick in and restart the pod, will make sure that the init containers are run in order, once both of them terminated successfully, our code will return `Ready`, and the member will be added again. Does that most edge cases?